dissect/volume/dm/thin.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import annotations
+
+from typing import BinaryIO, Optional
+
+from dissect.util.stream import AlignedStream
+
+from dissect.volume.dm.btree import BTree
+from dissect.volume.dm.c_dm import SECTOR_SIZE, c_dm
+from dissect.volume.exceptions import DMError
+
+
+class ThinPool:
+    """Interact with a device-mapper thin-provision pool."""
+
+    def __init__(self, metadata_fh: BinaryIO, data_fh: BinaryIO):
+        self.metadata_fh = metadata_fh
+        self.data_fh = data_fh
+
+        self.metadata = Metadata(metadata_fh)
+
+    def open(self, device_id: int, size_hint: Optional[int] = None) -> ThinDevice:
+        """Open a thin device on this pool.
+
+        No size information is stored in the pool, so it's recommended to provide a size hint.
+
+        Args:
+            device_id: The device ID to open.
+            size_hint: Optional size hint for the device to open.
+        """
+        return ThinDevice(self, device_id, size_hint)
+
+
+class Metadata:
+    """Interact with a device-mapper thin-provision metadata device."""
+
+    def __init__(self, fh: BinaryIO):
+        self.fh = fh
+        self.sb = c_dm.thin_disk_superblock(fh)
+        if self.sb.magic != c_dm.THIN_SUPERBLOCK_MAGIC:
+            raise DMError("Invalid magic for thin-pool metadata superblock")
+
+        self.data_block_size = self.sb.data_block_size
+        self.metadata_block_size = self.sb.metadata_block_size
+
+        self.data_mapping = BTree(self.fh, self.sb.data_mapping_root, self.metadata_block_size)
+        self.device_details = BTree(self.fh, self.sb.device_details_root, self.metadata_block_size)
+
+
+class ThinDevice(AlignedStream):
+    """Implements a readable thin device/volume.
+
+    Args:
+        pool: The pool the device belongs to.
+        device_id: The device ID of the device.
+        size_hint: Optional device size hint.
+    """
+
+    def __init__(self, pool: ThinPool, device_id: int, size_hint: Optional[int] = None):
```

Comment:
Makes more sense to me in this configuration, since a thin device is part of a pool. `pool.metadata` is the stable API for accessing metadata on a pool (it's not underscored).